Sampling of secondary photons from neutron collisions

samuelshaner opened the pull request:
This PR implements a routine for sampling secondary photons generated after neutron collisions using the same analog approach used in Serpent ([LINK](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&ved=0ahUKEwiR0MKlj-vUAhUBEz4KHVxzClsQFggoMAA&url=https%3A%2F%2Fwww.kns.org%2Fpaper_file%2Fpaper%2FMC2017_2017_2%2FP171S02-10LeppanenJ.pdf&usg=AFQjCNFthRPxaOQrCAEtQLhsDbPmCUixBQ)). A new `nu_photon_total` microscopic cross section is introduced that represents the summed total microscopic cross section times the photon yield for each reaction.  The procedure for sampling secondary photons is as follows:

    1) Following a collision, a nuclide is sampled.
    2) The average number of emitted photons is then given by the ratio of
       `nu_photon_total` to `total` for that nuclide.  An integral number of photons
        for the collision is sampled. 
    3) For each photon, the emission reaction is sampled.  
    4) The selected reaction product has an associated distribution used for sampling 
       the outgoing photon energy and angle from the incident neutron energy. 
    5) The outgoing direction is then sampled using the sampled angle.

Currently, OpenMC goes through the procedure of creating the secondary photons, but does not add them to the secondary bank because I was not able to generate the data needed for photon transport.

I should also note that the sampling of the secondary photons changes most of the test results as random numbers need to be used for photon sampling - should a new random number stream be added to sample secondary photons?

samuelshaner committed:
implemented sampling of secondary photons from neutron collisons

samuelshaner committed:
removed macroscopic nu_photon_total

samuelshaner committed:
removed unnecessary adjustment of indent

samuelshaner committed:
created script for adding photo data to cross_sections.xml and turned on photon transport

samuelshaner committed:
added sampling of electron and positron particle types

samuelshaner committed:
added more energy deposition tallies and worked on photoelectric rxn treatment

samuelshaner committed:
fixed issue with sampling distance to collison for electrons and positrons

samuelshaner committed:
turned on samping of secondary photons only if photon transport on

samuelshaner committed:
fixed issues with using a mesh filter and reading in U234 xs

samuelshaner committed:
fixed typo in library.py

samuelshaner commented:
I have made several updates to this PR:

- An `openmc-get-photo-endf71` script was added to download the photo-atomic and atomic relaxation ENDF files needed for photon transport. The `openmc-get-nndc-data` script and python API have been modified such that photon transport files can be downloaded during the download of NNDC data and the hdf5 files needed for photon transport can be added to an existing `cross_sections.xml` file.
- With the photon transport data able to be generated and read in, secondary photons (as well as any photons, electrons, or positrons they create) are now added in the secondary bank.
- An `electron_treatment` option has been added to use different models to treat electrons and positrons. The only model implemented is for the electrons and positrons to deposit all of their energy locally (i.e. don't produce any bremsstrahlung).  A constant for thick-target bremsstrahlung (TTB) has been added, but the physics for TTB have not been implemented.
- Tallies have been added for the energy deposited from fission fragments, betas, elastic scattering, photons, electrons, and positrons.  These tallies are pretty basic right now and haven't been thoroughly tested.

There are a number of items still to be resolved (though not necessarily all in this PR) that it would be great to get your thoughts on:

- It seems cleanest to have particle type implemented as a filter and only have some scores valid for certain particle types.
- Photon transport significantly increases run time due to the many daughter photons created by each photon produced from a neutron interaction. Variance reduction techniques should probably be made more nuanced to allow them to be applied to only specific particle types.
- Thick target bremsstrahlung has not been implemented. Haven't found a good reference that clearly outlines how to implement this. Looks like electro-atomic interaction data needs to be downloaded and read in. 
- Tests and further documentation need to be created.
- Test problem for benchmarking? Perhaps use the broomstick problem from the Serpent paper on implementing photon transport ([LINK](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&ved=0ahUKEwiR0MKlj-vUAhUBEz4KHVxzClsQFggoMAA&url=https%3A%2F%2Fwww.kns.org%2Fpaper_file%2Fpaper%2FMC2017_2017_2%2FP171S02-10LeppanenJ.pdf&usg=AFQjCNFthRPxaOQrCAEtQLhsDbPmCUixBQ))?

paulromano commented on openmc/data/library.py:
```diff
@@ -63,32 +63,29 @@ def register_file(self, filename):
-        common_dir = os.path.dirname(os.path.commonprefix(
-            [lib['path'] for lib in self.libraries]))
-        if common_dir == '':
-            common_dir = '.'
```
Was there a reason this was removed?

samuelshaner replied:
The `openmc-get-photo-endf71` script places the photo-atomic hdf5 files in a separate directory and appends their location to an existing `cross_sections.xml` file if provided with the -c argument.  If `cross_sections.xml` has a `directory` element,  then that must be taken into account when setting the location of the photo-atomic hdf5 files.  I figured instead of checking for a `directory` element and making the new hdf5 locations relative to that, just set the full location of all the hdf5 files.

paulromano commented on scripts/openmc-get-photo-endf71:
```diff
@@ -0,0 +1,93 @@
+
+library = openmc.data.DataLibrary()
+
+for z in range(1,101):
```
`for z in range(1, 101):`

paulromano commented on scripts/openmc-get-photo-endf71:
```diff
@@ -0,0 +1,93 @@
+                    help='cross_sections.xml file to append libraries to')
+args = parser.parse_args()
+
+base_url = 'http://www-nds.iaea.org/public/download-endf/ENDF-B-VII.1/'
```
I think it would be better to download zip files from NNDC and extract:
http://www.nndc.bnl.gov/endf/b7.1/zips/ENDF-B-VII.1-photoat.zip
http://www.nndc.bnl.gov/endf/b7.1/zips/ENDF-B-VII.1-atomic_relax.zip

samuelshaner replied:
Agreed, I didn't know there was a zip file. Nice find!

paulromano commented on src/input_xml.F90:
```diff
@@ -267,6 +267,34 @@ subroutine read_settings_xml()
+    if (check_for_node(root, "electron_treatment")) then
+      call get_node_value(root, "electron_treatment", temp_str)
+      select case (to_lower(temp_str))
+      case ("LED")
```
should be lower-case "led"

paulromano commented on src/input_xml.F90:
```diff
@@ -267,6 +267,34 @@ subroutine read_settings_xml()
+      select case (to_lower(temp_str))
+      case ("LED")
+        electron_treatment = ELECTRON_LED
+      case ("TTB")
```
should be lower-case "ttb"

paulromano commented on src/input_xml.F90:
```diff
@@ -267,6 +267,34 @@ subroutine read_settings_xml()
+      case default
+        call fatal_error("Unrecognized photon transport: " // &
+             trim(temp_str) // ".")
+      end select
```
You can just have
```Fortran
call get_node_value(root, "photon_transport", photon_transport)
```
And then get rid of the select case (i.e., it is ok to pass logicals to `get_node_value`)

paulromano commented on src/nuclide_header.F90:
```diff
@@ -117,6 +121,7 @@ module nuclide_header
     real(8) :: absorption      ! microscopic absorption xs
     real(8) :: fission         ! microscopic fission xs
     real(8) :: nu_fission      ! microscopic production xs
+    real(8) :: nu_photon_total ! microscopic photon production xs
```
I'm not crazy about the name `nu_photon_total` because people don't typically use the greek letter nu to talk about a yield of photons. I'd suggest something like `photon_prod` or even `photon_prod_xs` to indicate that it is a cross section and not a yield.

samuelshaner replied:
Neither am I. I'll change it to `photon_prod`.

paulromano commented on src/endf.F90:
```diff
@@ -66,6 +66,18 @@ pure function reaction_name(MT) result(string)
+      string = "fission-q-fragments"
+    case (SCORE_FISS_Q_BETAS)
+      string = "fission-q-betas"
+    case (SCORE_Q_ELASTIC)
```
I think we should change all these "Q" scores to something else since it is not really Q that we're tallying, it's energy deposition. It's a little more accurate to call the fission ones Q, but even there probably calling it energy release would be more obvious.

paulromano commented on src/tally.F90:
```diff
@@ -1173,6 +1230,166 @@ subroutine score_general_ce(p, t, start_index, filter_index, i_nuclide, &
+        ! Skip any non-elastic scatter events
+        if (p % event_MT /= ELASTIC) cycle SCORE_LOOP
+
+        score = p % wgt * (p % last_E - p % E)
```
We ought to think about the overall strategy for energy deposition. Since NJOY is hooked into `openmc.data` pretty nicely, we could take advantage of `HEATR` to get KERMA values which would then allow us to tally energy deposition using tracklength/collision estimators. I'll send you some papers about this and how MC21 uses those values to calculate energy deposition.

samuelshaner replied:
Agreed. I'm not attached to what I put in - I just wanted sometime to start tallying the energy deposition from different sources.  I look forward to reading those papers.

paulromano commented on src/photon_physics.F90:
```diff
@@ -305,13 +305,22 @@ recursive subroutine atomic_relaxation(p, elm, i_shell)
 
-    ! Check for no transitions
-    if (elm % shells(i_shell) % n_transitions == 0) return
+    ! If no transitions, assume fluorescent photon from captured free electron
```
What's the basis for this? Do other codes do this?

samuelshaner replied:
The basis for it is just maintaining energy balance.  If an electron hole is created (and especially one from an inner shell electron), the hole is definitely going to be filled.  If there is no shell to shell transition found, we still need to somehow fill it to ensure energy balance.  My thought was just to fill it with a free electron.

I'm not sure what other codes do.  It is in the weeds of handling atomic relaxation so I'm not sure if it will be reported in any theory manuals. I'll check the masters thesis you shared and try to find what is done in the Serpent source code.

paulromano commented on src/physics.F90:
```diff
@@ -422,6 +510,59 @@ subroutine sample_fission(i_nuclide, E, i_reaction)
+    type(Nuclide), pointer :: nuc
+
+    ! Get pointer to nuclide
+    nuc => nuclides(i_nuclide)
```
Use
```Fortran
associate (nuc => nuclides(i_nuclide))
```
and get rid of the nuclide pointer variable.

paulromano commented on src/physics.F90:
```diff
@@ -1498,4 +1639,50 @@ subroutine inelastic_scatter(nuc, rxn, p)
+
+      ! Sample the reaction and product
+      call sample_photon_product(i_nuclide, p % E, i_reaction, i_product)
+      rx => nuclides(i_nuclide) % reactions(i_reaction)
```
Use
```Fortran
associate (rx => nuclides(i_nuclide) % reactions(i_reaction))
```
and get rid of the reaction pointer variable.

paulromano commented on src/cross_section.F90:
```diff
@@ -563,6 +573,11 @@ subroutine calculate_urr_xs(i_nuclide, i_temp, E)
+      ! Set the nu-photon production cross section
+      i_grid = int(log(E/energy_min_neutron)/log_spacing)
+      micro_xs(i_nuclide) % nu_photon_total = &
+           nuc % compute_nu_photon_total(E, i_temp, i_grid)
```
Isn't the total photon production determined up-front? (i.e., what you added to sum_xs)

samuelshaner replied:
Yeah, that's right it doesn't need to be recomputed again when computing urr xs with probability tables.

paulromano commented on src/nuclide_header.F90:
```diff
@@ -539,6 +592,18 @@ subroutine nuclide_create_derived(this)
           this % sum_xs(t) % total(j:j+n-1) = this % sum_xs(t) % total(j:j+n-1) + &
                rx % xs(t) % value
 
+          ! Calculate nu-photon total cross section
```
We'll need to be careful because the photon production data is a bit messy for some evaluations. Often, you'll see photon production assigned to summed reactions like MT=4 or MT=3. Also, sometimes photon production is listed for reactions that don't have any corresponding cross sections. If you look at how we handle this in `openmc.data` you'll see that we have a `summed_reactions` dictionary that stores redundant reactions, some of which have photon production that doesn't appear elsewhere. We'll need to make sure that the total photon production that is listed in the ACE file is actually what we end up using. Anyway, for now you don't need to start digging into this too deeply but this is something we'll have to keep in mind as a TODO.

paulromano replied:
What we may end up needing to do is writing out those summed reactions to the HDF5 libraries.

paulromano commented on src/nuclide_header.F90:
```diff
@@ -690,6 +751,72 @@ pure function nuclide_nu(this, E, emission_mode, group) result(nu)
+      end associate
+    end do
+
+  end function compute_nuclide_nu_photon_total
```
Related to my other question -- isn't this function redundant since sum_xs already stores nu_photon_total?

samuelshaner replied:
Yeah, I'll remove this.

paulromano reviewed (changes requested):
Huge thanks for this first shot at implementing photon production! We'll need to make some changes and improvements, but this lays a lot of the groundwork. To respond to your specific questions:

- I agree that having a particle filter is the simplest way to handle tallies for multiple particles.
- One thing we might want to consider is having a 'photon source generation mode' or something like that that can then be used to run a fixed-source photon transport calc. This is what MC21 does actually.
- I haven't had a chance to read it yet, but this [master's thesis](https://aaltodoc.aalto.fi/bitstream/handle/123456789/21004/master_Kaltiaisenaho_Toni_2016.pdf?sequence=1) looks really awesome for details on photon transport in general. Might want to look there for details on TTB.
- A broomstick problem is probably a good starting point.

Regarding the earlier question about random number generation, it seems appropriate to have a separate stream for photons and it should be pretty trivial to implement.

samuelshaner committed:
added particle filter and addressed PR comments

samuelshaner committed:
fixed error in issuing particle filter warning

samuelshaner committed:
fixed assorted issues with heating scores

samuelshaner committed:
removed unnecessary PR edits

paulromano commented on src/constants.F90:
```diff
@@ -341,15 +341,21 @@ module constants
+       SCORE_Q_PHOTONS          = -32, & ! recoverable fission Q-value
+       SCORE_Q_ELECTRONS        = -33, & ! recoverable fission Q-value
+       SCORE_Q_POSITRONS        = -34, & ! recoverable fission Q-value
+       SCORE_Q_ELASTIC          = -35, & ! recoverable fission Q-value
```
Need to update the comments at the end of all of these

samuelshaner commented:
Thanks for the review!  I addressed most of your comments.  A few things to note:

- **Particle filter**: I added a particle filter.  While useful for things like heating, many scores don't make sense for particles other than neutrons or don't tally anything for particles other than neutrons. For instance all scattering tallies, only work for neutrons.
- **Energy deposition**: This will likely need to be significantly reworked.  Perhaps it is best to remove my additions and follow the procedure used in MC21?
- **Random number stream for photons**: I added a random number stream for photons called `STREAM_PHOTON`. It is used for tracking photons *and* sampling photon interaction reactors. The `STREAM_TRACKING` seems to encompass more than just tracking so I'm wondering if it would be clearer to have something more like `STREAM_GENERAL` and `STREAM_PHOTON`?

The thesis and papers you shared looked quite good.  I'll be reading those over today.

paulromano commented:
Yeah, maybe let's leave energy deposition alone until we decide what the overall strategy is.

samuelshaner committed:
added non-relativistic Sauter distribution for photoelectron outgoing angle

samuelshaner commented on src/physics.F90:
```diff
@@ -216,15 +231,73 @@ subroutine sample_photon_reaction(p)
+        call p % create_secondary(uvw, p % E, ELECTRON, run_CE=.true.)
+
+        ! TODO: figure out correct event_MT
+        p % event_MT = 533
```
The masters thesis on photon transport in Serpent says they give the whole photon energy to the electron if no shell was sampled. Do you have any recommendation on what `event_MT` to assign to this interaction?

paulromano replied:
I'm confused as to how this could happen. Isn't the total photoatomic cross section just the sum of the subshell cross sections??

samuelshaner replied:
It looks like the total and subshell photo atomic cross sections are read in separately. The total cross section is used to decide which interaction takes place (e.g. photoelectric, etc) and then the subshell cross sections are used to determine which electron gets ejected. 

Serpent 2 is only stores subshells cross section tables above a certain energy cut off, so the sum of the subshell cross sections is less than the stored total cross section, hence why this is needed. Since all subshell cross sections tables are stored in OpenMC, we shouldn't need this. There still could be a discrepancy due to the subshell xs not exactly summing to the total xs, but that *should* be small. Since OpenMC stores all subshell xs tables, I'll remove this.

samuelshaner committed:
removed option for giving photoelectron entire photon energy

samuelshaner committed:
removed new energy deposition scores
